_posts/2019-04-05-omakase-typescript.markdown

```diff
@@ -0,0 +1,179 @@
+is a good primer if you want to learn more.)
+
+Both of these variables contain a string, but the difference is that in Java the variable can _only_ ever contain a
+_string_, because that's what we said when we created the variable. In JS the variable can change to be _anything_,
+like a number, or a list of dates.
+
+To illustrate:
+
+```js
+// Before in JS
+var name = "Danger"
+// Also OK
+var name = 1
+var name = false
+var name = ["2018-02-03", "2019-01-12"]
+
+// Before in Java
+String name = "Danger";
+// Not OK, the code wouldn't be accepted by Java
+String name = 1;
+String name = false
+String name = new String[]{"2018-02-03", "2019-01-12"};
+```
+
+These trade-offs make sense in the context for which these languages were built back in 1995. JavaScript was
+originally designed to be a small programming language which handled simple interactions on websites. Java on the
+other hand was built specifically to make big apps which could run on any computer. Their needs had different
+scales, so the language required programmers write different types of code.
+
+Java required programmers to be more explicit with the values of their variables because the programs they expected
+people to build were more complex. While JavaScript opted for ease of reading, and flexibility because you would be
+able to see most of the code in one go.
+
+### What is TypeScript?
+
+TypeScript is a programming language - it contains all of JavaScript, and then a bit more. Using our example above,
+let's compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:
+
+```js
+// JavaScript
+var name = "Danger"
+console.log("Hello, " + danger)
+
+// TypeScript
+var name = "Danger"
+console.log("Hello, " + danger)
+
+// Yep, you're not missing something, there's no difference
+```
+
+Due to TypeScript's aim to only _extend_ JavaScript, your normal JavaScript code should work fine with TypeScript.
+The things TypeScript adds to JavaScript are intended to help you be more explicit about what kinds of data are
+used in your code, a bit like Java.
+
+```diff
+- var name = "Danger"
++ var name: string = "Danger"
+console.log("Hello, " + danger)
+```
+
+This extra `: string` allows the reader to be certain that `name` will only be a string. Annotating your variables
+also allows TypeScript to check this for you. This is very useful because keeping track of changes like the type of
+value in a variable seems easy when it's one or two, but once it starts hitting the hundreds, that's a lot to keep
+track of.
+
+Simply speaking, we call these annotations "Types". Hence the name <i>Type</i>Script. The tag-line for TypeScript
+is "JavaScript which scales" which is a statement that these extra type annotations allows you to work on bigger
+projects. This is because you can verify up-front how correct your code is. This means you have less need to
+understand how every change affects the rest of the program.
+
```

Comment:
This section could use a summarizing paragraph. It's not clear why TypeScript is appropriate _now_, while JavaScript not having types made sense in the 90's. Throughout this post, we've contrasted the two languages, which is a lot to grok, and we should reward the reader with an understanding of why the tradeoffs worked out the way they did.

Reply:
perfect, this was a killer point!